windows.c

```diff
@@ -193,15 +193,30 @@ int doToggleProxy(bool turnOn)
   return ret;
 }
 
+HWND hWnd = NULL;
+HANDLE hInvisibleThread = NULL;
+
 int toggleProxy(bool turnOn)
 {
-  return doToggleProxy(turnOn);
+  if (hInvisibleThread == NULL) {
+    return doToggleProxy(turnOn);
+  }
+
+  // in this case, we're running in wait-and-cleanup mode, close the Window and
+  // let that trigger the toggling on the event loop thread.
+  PostMessage(hWnd, WM_CLOSE, 0, 0);
+  WaitForSingleObject(hInvisibleThread, INFINITE);
+  DWORD exitCode = 0;
+  GetExitCodeThread(hInvisibleThread, &exitCode);
+  return exitCode;
```

Comment:
The basic idea is that when we're running with `wait-and-cleanup`, we should always to the toggling on the event loop thread. This way, we avoid race conditions between cleanup due to stdin getting closed vs cleanup due to system shutting down.

Reply:
Nice, will try

Reply:
You can use this sysproxy branch - https://github.com/getlantern/sysproxy/pull/10

Reply:
https://github.com/getlantern/sysproxy-cmd/pull/11 shows that our code calls the API correctly but the OS doesn't actually clear the proxy setting. It's unclear why but might relate to the VM software.

The action is to set proxy off in both cases, so race condition shouldn't cause trouble in theory.